src/stream/src/executor/temporal_filter.rs

```diff
@@ -0,0 +1,466 @@
+                            self.state_table.delete(row);
+                        }
+                    }
+                }
+            } else {
+                new_ops.push(op);
+                new_visibility.append(false);
+            }
+        }
+
+        let new_visibility = new_visibility.finish();
+
+        (new_ops, new_visibility)
+    }
+
+    /// Fill the buffer when the temporal executor initializes or scales. On initialization
+    /// (including recovering), `prev_vnode_bitmap` will be `None`. On scaling,
+    /// `prev_vnode_bitmap` should be set as the previous vnode bitmap to perform buffer update.
+    /// Note that we do not assume set relations between `prev_vnode_bitmap` and
+    /// `curr_vnode_bitmap` when scaling occurs. That is to say, `prev_vnode_bitmap` does not
+    /// necessarily contain `curr_vnode_bitmap` on scaling out, and vice versa. Therefore, we
+    /// always check vnodes that are no longer owned and newly owned by the temporal executor
+    /// regardless of the scaling type (scale in or scale out).
+    async fn fill_buffer(
+        &mut self,
+        prev_vnode_bitmap: Option<&Bitmap>,
+        curr_vnode_bitmap: &Bitmap,
+    ) -> StreamExecutorResult<()> {
+        // When scaling occurs, we remove data with vnodes that are no longer owned by this executor
+        // from buffer.
+        if let Some(prev_vnode_bitmap) = prev_vnode_bitmap {
+            let no_longer_owned_vnodes =
+                Bitmap::bit_saturate_subtract(prev_vnode_bitmap, curr_vnode_bitmap);
+            self.buffer.retain(|(_, pk), _| {
+                let vnode = self.state_table.compute_vnode(pk);
+                !no_longer_owned_vnodes.is_set(vnode as _)
+            });
+        }
+
+        let values_per_vnode =
+            get_rowstreams_by_new_vnodes(prev_vnode_bitmap, curr_vnode_bitmap, &self.state_table)
+                .await?;
+        if !values_per_vnode.is_empty() {
+            let mut stream = select_all(values_per_vnode);
+            while let Some(storage_result) = stream.next().await {
+                // Insert the data into buffer.
+                let row = storage_result?.into_owned();
+                let timestamp = row[self.time_col_idx].as_ref().unwrap().clone();
+                let pk = row.by_indices(&self.pk_indices);
+                self.buffer.insert((timestamp, pk), row);
+            }
+        }
+        Ok(())
+    }
+
+    fn sync(
+        &mut self,
+        lower_condition: &BoxedExpression,
+    ) -> StreamExecutorResult<Vec<StreamChunk>> {
+        let identity = self.identity.clone();
+        let mut ret = vec![];
+        let mut data_chunk_builder =
+            DataChunkBuilder::new(self.schema().data_types(), self.chunk_size);
+        while let Some(entry) = self.buffer.first_entry() {
+            if lower_condition
+                .eval_row_infallible(entry.get(), |err| self.ctx.on_compute_error(err, &identity))
```

Comment:
Anw, this seems very inefficient. It may be preferable to evict an entire range.

Reply:
> Anw, this seems very inefficient. It may be preferable to evict/forward an entire range.

same as below